File: carbon/src/main/java/carbon/view/InputView.java
package carbon.view;

import androidx.annotation.NonNull;
import carbon.widget.OnValidateListener;

/**
 * Interface of a view with input, which can be validated. Used by {@link
 * carbon.widget.InputLayout}
 */
public interface InputView extends ValidStateView {
    /**
     * Performs validation
     */
    void validate();

    /**
     * Adds a listener
     *
     * @param listener cannot be null
     */
    void addOnValidateListener(@NonNull OnValidateListener listener);

    /**
     * Removes a listener
     *
     * @param listener cannot be null
     */
    void removeOnValidateListener(@NonNull OnValidateListener listener);

    /**
     * Removes all listeners
     */
    void clearOnValidateListeners();
}

File: carbon/src/main/java/carbon/view/InputView.kt
package carbon.view

import carbon.widget.OnValidChangedListener
import carbon.widget.OnValidateListener

/**
 * Interface of a view with input, which can be validated. Used by [ ]
 */
interface InputView : ValidStateView {
    /**
     * Performs validation
     */
    fun validate()

    /**
     * Adds a listener
     *
     * @param listener cannot be null
     */
    fun addOnValidateListener(listener: OnValidateListener)

    /**
     * Removes a listener
     *
     * @param listener cannot be null
     */
    fun removeOnValidateListener(listener: OnValidateListener)

    /**
     * Removes all listeners
     */
    fun clearOnValidateListeners()

    /**
     * Adds a listener
     *
     * @param listener cannot be null
     */
    fun addOnValidChangedListener(listener: OnValidChangedListener)

    /**
     * Removes a listener
     *
     * @param listener cannot be null
     */
    fun removeOnValidChangedListener(listener: OnValidChangedListener)

    /**
     * Removes all listeners
     */
    fun clearOnValidChangedListeners()
}
